lib/FtpContext.js

```diff
@@ -279,39 +332,56 @@ module.exports = class FTPContext {
      * or a socket event, like an error or timeout.
      *
      * @private
-     * @param {Object} response
+     * @param {(Error | undefined)} err
+     * @param {(FTPResponse | undefined)} [response]
      */
-    _passToHandler(response) {
-        if (this._handler) {
-            this._handler(response, this._task);
+    _passToHandler(err, response) {
+        if (this._task) {
+            this._task.responseHandler(err, response, this._task.resolver);
         }
+        // Errors other than FTPError always close the client. If there isn't an active task to handle the error,
+        // the next one submitted will receive it using `_closingError`.
+        // There is only one edge-case: If there is an FTPError while no task is active, the error will be dropped.
+        // But that means that the user sent an FTP command with no intention of handling the result. So why should the
+        // error be handled? Maybe log it at least? Debug logging will already do that and the client stays useable after
+        // FTPError. So maybe no need to do anything here.
     }
 
     /**
-     * Reset the state of this context.
+     * Send an error to the current handler and close all connections.
      *
      * @private
+     * @param {Error} err
      */
-    _reset() {
-        this.log("Closing connections.");
-        this._stopTrackingTask();
-        this._partialResponse = "";
+    _closeWithError(err) {
+        this._closingError = err;
+        // Before giving the user's task a chance to react, make sure we won't be bothered with any inputs.
```

Comment:
They'll be garbage collected eventually. I definitely don't want to do this before calling the task handler because I want to give task code as much freedom as possible. Maybe set them to undefined (also ok?) later…